src/gpu/matrix/KmMatrix/Arith.cu

```diff
@@ -0,0 +1,169 @@
+#include "Arith.hpp"
+#include "../../utils/GpuInfo.cuh"
+
+namespace h2o4gpu {
+namespace Matrix {
+
+namespace kernel {
+
+/*
+ * Compute min value for each row.
+ * @tparam T Numeric type of the data
+ * @param _res The output matrix with shape m x 1
+ * @param _val The input matrix with shape m x n
+ */
+template <typename T>
+__global__ void row_min_sequential(kParam<T> _res, kParam<T> _val) {
```

Comment:
These two kernels seem to be duplicated in src/gpu/kmeans/kmeans_init.cu. Did you move some code around and forget to remove it?

Reply:
Not exactly. These two deals with min and argmin, is my original attempt. But then the 0 distance bug surfaced, I created two similar kernels specifically dealing with 0 values in kmeans_init.cu. The former is not used in kmeans, the latter is not generic enough to be put in matrix. I will look into cub function you suggested, hopefully we remove the latter two.